### PR TITLE
Fix batch_put ttl issue

### PR DESCRIPTION
**Pull request by limbooverlambda**
There is a bug in the client where the ttls are not properly sharded for RawBatchPuts. This essentially means that if the batch is spread across multiple regions, the ttls for the batch specific to the shard is not properly aligned. This is an issue when the request ends up in the TiKV node specific to the region and fails because of the misalignment of the `Vec<TTLs`> with its corresponding `Vec<KvPair>`. The fix was to change the sharding logic for RawBatchPutRequest so that the TTLs are aligned with the KvPair. Maintaining a mapping of the kvpair to its corresponding ttl in the Shard output.

- Added unit tests for the batch_put
- Ran integ tests to confirm:

```
warning: `tikv-client` (lib) generated 4 warnings (run `cargo fix --lib -p tikv-client` to apply 2 suggestions)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.61s
     Running tests/integration_tests.rs (target/debug/deps/integration_tests-ae58f87a6b1272e6)

running 26 tests
test raw_bank_transfer ... ok
test raw_req ... ok
test txn_crud ... ok
test txn_batch_mutate_optimistic ... ok
test raw_ttl ... ok
test raw_cas ... ok
test txn_bank_transfer ... ok
test txn_batch_mutate_pessimistic ... ok
test txn_get_for_update ... ok
test raw_write_million ... ok
test txn_get_timestamp ... ok
test txn_insert_duplicate_keys ... ok
test txn_key_exists ... ok
test txn_lock_keys ... ok
test txn_lock_keys_error_handle ... ok
test txn_pessimistic ... ok
test txn_pessimistic_delete ... ok
test txn_pessimistic_heartbeat ... ok
test txn_pessimistic_rollback ... ok
test txn_read ... ok
test txn_scan ... ok
test txn_scan_reverse ... ok
test txn_scan_reverse_multi_regions ... ok
test txn_split_batch ... ok
test txn_unsafe_destroy_range ... ok
test txn_update_safepoint ... ok

test result: ok. 26 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 60.74s


```

### CODE REVIEW

File: src/raw/requests.rs [hunk 1]
```diff
@@ -190,23 +192,44 @@ impl KvRequest for kvrpcpb::RawBatchPutRequest {
...
+    type Shard = Vec<(kvrpcpb::KvPair, u64)>;
 
     fn shards(
         &self,
         pd_client: &Arc<impl PdClient>,
     ) -> BoxStream<'static, Result<(Self::Shard, RegionStore)>> {
+        // Maintain a map of the pair and its associated ttl
+        let kvs = self.pairs.clone();
+        let kv_pair = kvs.into_iter().map(Into::<KvPair>::into);
+        let kv_ttl = kv_pair.zip(self.ttls.clone()).collect::<HashMap<_, _>>();
```
Comment (pingyu) [inline src/raw/requests.rs:204]: `store_stream_for_keys` is able to handle data of keys along with values and ttls. Mapping by Hash map is not necessary, and it's not efficient as well.

Please refer to [`PrewriteRequest::shards`](https://github.com/tikv/client-rust/blob/54fd72001bca74288d02134809424aeea7d3f19d/src/transaction/requests.rs#L290).

Comment (limbooverlambda) [reply src/raw/requests.rs]: Thanks for the review. Didn't realize `store_stream_for_keys` has this capability. Removed the HashMap to collect the ttls for the kv_pair.

Comment (limbooverlambda) [reply src/raw/requests.rs]: Hi @pingyu, I have made the changes. Please let me know if there's anything else I need to consider for this PR.

File: src/kv/key.rs [hunk 2]
```diff
@@ -79,6 +80,20 @@ impl AsRef<Key> for kvrpcpb::Mutation {
...
+
+impl AsRef<Key> for KvPairTTL {
+    fn as_ref(&self) -> &Key {
+        self.0.key.as_ref()
+    }
+}
+
+impl From<KvPairTTL> for (KvPair, u64) {
+    fn from(value: KvPairTTL) -> Self {
+        (value.0.clone(), value.1)
```
Comment (pingyu) [inline src/kv/key.rs:93]:
```suggestion
        (value.0, value.1)
```

File: src/raw/requests.rs [hunk 3]
```diff
@@ -190,23 +193,28 @@ impl KvRequest for kvrpcpb::RawBatchPutRequest {
...
-        let mut pairs = self.pairs.clone();
-        pairs.sort_by(|a, b| a.key.cmp(&b.key));
-        store_stream_for_keys(
-            pairs.into_iter().map(Into::<KvPair>::into),
-            pd_client.clone(),
-        )
+        let kvs = self.pairs.clone();
+        let ttls = self.ttls.clone();
+        let mut kv_ttl: Vec<KvPairTTL> = kvs
+            .iter()
```
Comment (pingyu) [inline src/raw/requests.rs:205]:
```suggestion
            .into_iter()
```

File: src/raw/requests.rs [hunk 4]
```diff
@@ -190,23 +193,28 @@ impl KvRequest for kvrpcpb::RawBatchPutRequest {
...
+            .zip(ttls)
+            .map(|(kv, ttl)| KvPairTTL(kv.clone(), ttl))
```
Comment (pingyu) [inline src/raw/requests.rs:207]:
```suggestion
            .map(|(kv, ttl)| KvPairTTL(kv, ttl))
```

File: src/raw/requests.rs [hunk 5]
```diff
@@ -190,23 +193,28 @@ impl KvRequest for kvrpcpb::RawBatchPutRequest {
...
+            .collect();
+        kv_ttl.sort_by(|a, b| a.0.key.clone().cmp(&b.0.key));
+        store_stream_for_keys(kv_ttl.into_iter(), pd_client.clone()).boxed()
```
Comment (pingyu) [inline src/raw/requests.rs:210]:
```suggestion
        store_stream_for_keys(kv_ttl.into_iter(), pd_client.clone())
```

File: src/raw/requests.rs [hunk 6]
```diff
@@ -600,4 +621,59 @@ mod test {
...
+
+        let batch_put_request =
+            new_raw_batch_put_request(pairs.clone(), ttls.clone(), Some(cf), false);
+        let keyspace = Keyspace::Enable { keyspace_id: 0 };
+        let plan = crate::request::PlanBuilder::new(client, keyspace, batch_put_request)
+            .resolve_lock(OPTIMISTIC_BACKOFF, keyspace)
+            .retry_multi_region(DEFAULT_REGION_BACKOFF)
+            .plan();
+        let _ = plan.execute().await;
+        assert_eq!(actual_map.lock().unwrap().len(), 2);
```
Comment (pingyu) [inline src/raw/requests.rs:675]:
```suggestion
```

If the next `assert_eq` is passed, the length must be equal.

File: src/raw/requests.rs [hunk 7]
```diff
@@ -600,4 +621,59 @@ mod test {
...
+        assert_eq!(actual_map.lock().unwrap().deref().clone(), expected_map);
```
Comment (pingyu) [inline src/raw/requests.rs:676]:
```suggestion
        assert_eq!(actual_map.lock().unwrap().deref(), &expected_map);
```

File: src/store/mod.rs [hunk 8]
```diff
@@ -36,7 +36,7 @@ pub struct RegionStore {
 pub struct Store {
     pub client: Arc<dyn KvClient + Send + Sync>,
 }
-
+#[allow(dead_code)]
```
Comment (pingyu) [inline src/store/mod.rs:39]:
```suggestion
```

**Review by pingyu — COMMENTED** [review]
Rest LGTM.

File: src/kv/key.rs (hunk 2 above)
Comment (limbooverlambda) [reply src/kv/key.rs]: Changed

File: src/raw/requests.rs (hunk 3 above)
Comment (limbooverlambda) [reply src/raw/requests.rs]: Changed

File: src/raw/requests.rs (hunk 4 above)
Comment (limbooverlambda) [reply src/raw/requests.rs]: Changed

File: src/raw/requests.rs (hunk 5 above)
Comment (limbooverlambda) [reply src/raw/requests.rs]: Done

File: src/raw/requests.rs (hunk 6 above)
Comment (limbooverlambda) [reply src/raw/requests.rs]: Yup! Removed

File: src/raw/requests.rs (hunk 7 above)
Comment (limbooverlambda) [reply src/raw/requests.rs]: Changed.

File: src/store/mod.rs (hunk 8 above)
Comment (limbooverlambda) [reply src/store/mod.rs]: Removed

**Comment by limbooverlambda** [comment]
Thanks for helping me clean up this PR @pingyu. Addressed your comments.

File: src/raw/requests.rs [hunk 9]
```diff
@@ -7,10 +7,12 @@ use std::time::Duration;
 
 use async_trait::async_trait;
 use futures::stream::BoxStream;
+
```
Comment (pingyu) [inline src/raw/requests.rs:10]:
```suggestion
```

File: src/raw/requests.rs [hunk 10]
```diff
@@ -190,23 +192,28 @@ impl KvRequest for kvrpcpb::RawBatchPutRequest {
...
-            pd_client.clone(),
-        )
+        let kvs = self.pairs.clone();
+        let ttls = self.ttls.clone();
+        let mut kv_ttl: Vec<KvPairTTL> = kvs
+            .into_iter()
+            .zip(ttls)
+            .map(|(kv, ttl)| KvPairTTL(kv, ttl))
+            .collect();
+        kv_ttl.sort_by(|a, b| a.0.key.clone().cmp(&b.0.key));
```
Comment (pingyu) [inline src/raw/requests.rs:208]:
```suggestion
        kv_ttl.sort_by(|a, b| a.0.key.cmp(&b.0.key));
```

**Review by pingyu — COMMENTED** [review]
I'm sorry I missed another suggestion in last comments. The `clone` in comparation should also be un-necessary.

File: src/raw/requests.rs (hunk 9 above)
Comment (limbooverlambda) [reply src/raw/requests.rs]: done.

File: src/raw/requests.rs (hunk 10 above)
Comment (limbooverlambda) [reply src/raw/requests.rs]: Removed.

**Comment by limbooverlambda** [comment]
> I'm sorry I missed another suggestion in last comments. The `clone` in comparation should also be un-necessary.

Yeah, I should have caught that. Removed.

**Review by pingyu — APPROVED** [review]
LGTM~
Thanks for you contribution !

**Pull request merged**